### PR TITLE
C++ API Documentation: Exclude files in src/, fix issue with excludes and add option to specify custom excludes

### PR DESCRIPTION
**Pull request by luator**
## Description

While working on the documentation of a package, I noticed that also classes/functions defined in cpp files are included.  Since they are not part of the public API, I think they should also not appear in the HTML documentation so I added `src/` (and also `test/`) to the exclude patterns.

Further prepended all the exclude patterns with the package path. This is important as otherwise the patterns can also match with higher-up parts of the paths. E.g. when adding only src/*, this results in everything being removed as all packages are in .../workspace/src/.... I belive that this will also fix the long-existing issue that C++ API is not included in the documentation builds on Bamboo: On Bamboo the workspace is in a path like /home/bambooagent/bamboo-agent-home/xml-data/build-dir/MGC-SHAREDM-U2/ and the */*build*/* exclude pattern (which I removed now) was matching with the build-dir-part, thus excluding all files in the workspace...

To be able to further limit the input, an argument `DOXYGEN_EXCLUDE_PATTERNS` is added to the `add_[sphinx_]documentation()` macro.  If set, the value is simply added to `EXCLUDE_PATTERNS` in the doxygen config file.
My specific use-case for this is that I have a third-party header-only library included in the package, which I don't want to appear in the documentation.

## How I Tested

By building documentation of the real_time_tools package.

## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [x] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [x] All new functions/classes are documented and existing documentation is updated according to changes.
- [x] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

### CODE REVIEW

**Comment by MaximilienNaveau** [comment]
You also need to keep the demos folder to get the code snippet for documentation purposes.

**Comment by luator** [comment]
What ist the expected behaviour with the demos?  With the current master, there don't seem to be any code snippets from the demos appearing in the documentation.  I just get the files listed like this, which isn't very useful in my opinion (example from the shared_memory package):

image

**Comment by MaximilienNaveau** [comment]
Look at the end of this:

https://machines-in-motion.github.io/code_documentation/real_time_tools/documentation/doxygen_index_one_page.html

**Comment by luator** [comment]
I couldn't find a really nice solution so what I did now is to drop the commits for restricting to headers in `include/` and instead add `src/` (and also `test/`) to the excludes.  I tested with the real_time_tools package and the result looks good to me (demos are still included but it is not cluttered with all the stuff from src/ anymore).

I also prepended all the exclude patterns with the package path.  This is important as otherwise the patterns can also match with higher-up parts of the paths.  E.g. when adding only `src/*`, this results in everything being removed as all packages are in `.../workspace/src/...`.  I belive that this will also fix the long-existing issue that C++ API is not included in the documentation builds on Bamboo:  On Bamboo the workspace is in a path like `/home/bambooagent/bamboo-agent-home/xml-data/build-dir/MGC-SHAREDM-U2/` and the `*/*build*/*` exclude pattern (which I removed now) was matching with the `build-dir`-part, thus excluding all files in the workspace...

**Comment by luator** [comment]
@MaximilienNaveau @vincentberenz Did you have a look at my latest change already?  I'd like to get this merged to see if it really fixes the issue on Bamboo :)

File: cmake/sphinx.cmake [hunk 1]
```diff
@@ -169,8 +169,20 @@ endmacro()
 #   Please refer to the *Sphinx* paragraph in the *General Documentation* in this
 #   package for more explanation about the parametrization of the tools.
 #
+#   Optional Arguments:
+#
+#   - DOXYGEN_EXCLUDE_PATTERNS:  String that is added to EXCLUDE_PATTERNS in the
```
Comment (vincentberenz) [inline cmake/sphinx.cmake:174]: what format should this string have ?

**Review by vincentberenz — CHANGES_REQUESTED** [review]

**Review by MaximilienNaveau — APPROVED** [review]

File: cmake/sphinx.cmake (hunk 1 above)
Comment (luator) [reply cmake/sphinx.cmake:174]: I added more explanation with an example.

**Comment by luator** [comment]
I made some small change to allow multiple exclude patterns in the argument

**Review by vincentberenz — APPROVED** [review]

**Comment by luator** [comment]
Good news, this indeed fixed the issue on Bamboo :tada:

**Pull request merged**